homeassistant/components/pglab/__init__.py

```diff
@@ -0,0 +1,87 @@
+"""PG LAB Electronics integration."""
+from __future__ import annotations
+
+from pypglab.mqtt import Client, Sub_State, Subcribe_CallBack
+
+from homeassistant.components import mqtt
+from homeassistant.components.mqtt import ReceiveMessage
+from homeassistant.components.mqtt.subscription import (
+    async_prepare_subscribe_topics,
+    async_subscribe_topics,
+    async_unsubscribe_topics,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+
+from .const import (
+    DEVICE_ALREADY_DISCOVERED,
+    DISCONNECT_COMPONENT,
+    DISCOVERY_INSTANCE,
+    PLATFORMS,
+)
+from .discovery import CreateDiscovery
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up PG LAB Electronics integration from a config entry."""
+
+    # define the call back for pglab  module to publish a mqtt message
+    async def mqtt_publish(
+        topic: str, payload: str, qos: int | None = 0, retain: bool | None = False
+    ) -> None:
+        await mqtt.async_publish(hass, topic, payload, qos, retain)
+
+    # define the call back for pglab module to subscribe to a mqtt message
+    async def mqtt_subscribe(
+        sub_state: Sub_State, topic: str, callback_func: Subcribe_CallBack
+    ) -> Sub_State:
+        async def discovery_message_received(msg: ReceiveMessage) -> None:
```

Comment:
Why is this a coroutine function?

Reply:
Please see previous comment.

Reply:
As requested I'm moving the callback to be global.
I'm also change the name in mqtt_subscribe_callback.

Reply:
The question was why `mqtt_message_received_callback` is a coroutine function

Reply:
`mqtt_message_received_callback` has been renamed to `discovery_message_received` and is no longer a coroutine function, this issue is resolved 👍